### PR TITLE
[SYCLomatic] Add mapping query for 20 CUB APIs

### PR DESCRIPTION
**Pull request by yihwang-nv**

### CODE REVIEW

File: clang/examples/DPCT/CUB/cub$$DeviceScan$$InclusiveScanByKey.cu [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    return lhs == rhs;
+  }
+};
+
+void test(void *temp_storage, size_t &temp_storage_bytes, int *d_keys_in, int * d_values_in,int *d_values_out,
+          CustomSum scan_op, int init_value,int num_items, CustomEqual equality_op, cudaStream_t stream) {
+  // Start
+  cub::DeviceScan::InclusiveScanByKey(temp_storage/*void **/, temp_storage_bytes/*size_t &*/, d_keys_in/*KeysInputIteratorT*/, d_values_in/*ValuesInputIteratorT*/, d_values_out/*ValuesOutputIteratorT*/, scan_op/*ScanOpT*/, num_items/*int*/, equality_op/*EqualityOpT*/, stream/*cudaStream_t*/);
+  // End
+}
```
Comment (tangjj11) [inline clang/examples/DPCT/CUB/cub$$DeviceScan$$InclusiveScanByKey.cu]: no new line

File: clang/test/dpct/query_api_mapping/CUB/cub_device.cu [hunk 2]
```diff
@@ -0,0 +1,119 @@
+// RUN: dpct --cuda-include-path="%cuda-path/include" --stop-on-parse-err --query-api-mapping=cub::DeviceReduce::ArgMax | FileCheck %s -check-prefix=CHECK_REDUCE_ARG_MAX
+// CHECK_REDUCE_ARG_MAX: CUDA API:
+// CHECK_REDUCE_ARG_MAX-NEXT:  cub::DeviceReduce::ArgMax(temp_storage/*void **/, temp_storage_bytes/*size_t &*/, d_in/*InputIteratorT*/, d_out/*OutputIteratorT*/, num_items/*int*/, stream/*cudaStream_t*/);
+// CHECK_REDUCE_ARG_MAX-NEXT:Is migrated to:
+// CHECK_REDUCE_ARG_MAX-NEXT:  dpct::reduce_argmax(oneapi::dpl::execution::device_policy(*stream), d_in, d_out, num_items);
```
Comment (tangjj11) [inline clang/test/dpct/query_api_mapping/CUB/cub_device.cu:13]: need add like other lit test:
// CHECK_REDUCE_ARG_MAX-EMPTY:

**Review by tangjj11 — CHANGES_REQUESTED** [review]

File: clang/examples/DPCT/CUB/cub$$DeviceScan$$InclusiveScanByKey.cu (hunk 1 above)
Comment (yihwang-nv) [reply clang/examples/DPCT/CUB/cub$$DeviceScan$$InclusiveScanByKey.cu]: fixed

**Review by wenhuiNi — APPROVED** [review]

**Review by tangjj11 — APPROVED** [review]

File: clang/test/dpct/query_api_mapping/CUB/cub_device.cu (hunk 2 above)
Comment (yihwang-nv) [reply clang/test/dpct/query_api_mapping/CUB/cub_device.cu:13]: fixed

File: clang/test/dpct/query_api_mapping/CUB/cub_device.cu [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+// CHECK_REDUCE_ARG_MIN-NEXT:   cub::DeviceReduce::ArgMin(temp_storage/*void **/, temp_storage_bytes/*size_t &*/, d_in/*InputIteratorT*/, d_out/*OutputIteratorT*/, num_items/*int*/, stream/*cudaStream_t*/);
+// CHECK_REDUCE_ARG_MIN-NEXT: Is migrated to:
+// CHECK_REDUCE_ARG_MIN-NEXT:   dpct::reduce_argmin(oneapi::dpl::execution::device_policy(*stream), d_in, d_out, num_items);
+// CHECK_REDUCE_ARG_MIN-EMPTY:
+
+// RUN: dpct --cuda-include-path="%cuda-path/include" --stop-on-parse-err --query-api-mapping=cub::DeviceReduce::Max | FileCheck %s -check-prefix=CHECK_REDUCE_MAX
+// CHECK_REDUCE_MAX: CUDA API:
+// CHECK_REDUCE_MAX-NEXT:   cub::DeviceReduce::Max(temp_storage/*void **/, temp_storage_bytes/*size_t &*/, d_in/*InputIteratorT*/, d_out/*OutputIteratorT*/, num_items/*int*/, stream/*cudaStream_t*/);
+// CHECK_REDUCE_MAX-NEXT: Is migrated to:
+// CHECK_REDUCE_MAX-NEXT:   stream->fill(d_out, oneapi::dpl::reduce(oneapi::dpl::execution::device_policy(*stream), d_in, d_in + num_items, typename std::iterator_traits<decltype(d_out)>::value_type{}, sycl::maximum<>()), 1).wait();
```
Comment (zhimingwang36) [inline clang/test/dpct/query_api_mapping/CUB/cub_device.cu:35]: Please declare a stream variable.

File: clang/test/dpct/query_api_mapping/CUB/cub_device.cu [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+// CHECK_REDUCE_MAX-EMPTY:
+
+// RUN: dpct --cuda-include-path="%cuda-path/include" --stop-on-parse-err --query-api-mapping=cub::DeviceReduce::Min | FileCheck %s -check-prefix=CHECK_REDUCE_MIN
+// CHECK_REDUCE_MIN: CUDA API:
+// CHECK_REDUCE_MIN-NEXT:   cub::DeviceReduce::Min(temp_storage/*void **/, temp_storage_bytes/*size_t &*/, d_in/*InputIteratorT*/, d_out/*OutputIteratorT*/, num_items/*int*/, stream/*cudaStream_t*/);
+// CHECK_REDUCE_MIN-NEXT: Is migrated to:
+// CHECK_REDUCE_MIN-NEXT:   stream->fill(d_out, oneapi::dpl::reduce(oneapi::dpl::execution::device_policy(*stream), d_in, d_in + num_items, typename std::iterator_traits<decltype(d_out)>::value_type{}, sycl::minimum<>()), 1).wait();
```
Comment (zhimingwang36) [inline clang/test/dpct/query_api_mapping/CUB/cub_device.cu:46]: same as above.

File: clang/test/dpct/query_api_mapping/CUB/cub_device.cu [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+// CHECK_REDUCE_BY_KEY-NEXT: Is migrated to:
+// CHECK_REDUCE_BY_KEY-NEXT:   stream->fill(d_num_runs_out, std::distance(d_unique_out, oneapi::dpl::reduce_by_key(oneapi::dpl::execution::device_policy(*stream), d_keys_in, d_keys_in + num_items, d_values_in, d_unique_out, d_aggregates_out, std::equal_to<typename std::iterator_traits<decltype(d_keys_in)>::value_type>(), op).first), 1).wait();
+// CHECK_REDUCE_BY_KEY-EMPTY:
+
+// RUN: dpct --cuda-include-path="%cuda-path/include" --stop-on-parse-err --query-api-mapping=cub::DeviceReduce::Sum | FileCheck %s -check-prefix=CHECK_REDUCE_SUM
+// CHECK_REDUCE_SUM: CUDA API:
+// CHECK_REDUCE_SUM-NEXT:   cub::DeviceReduce::Sum(temp_storage/*void **/, temp_storage_bytes/*size_t &*/, d_in/*InputIteratorT*/, d_out/*OutputIteratorT*/, num_items/*int*/, stream/*cudaStream_t*/);
+// CHECK_REDUCE_SUM-NEXT: Is migrated to:
+// CHECK_REDUCE_SUM-NEXT:   stream->fill(d_out, oneapi::dpl::reduce(oneapi::dpl::execution::device_policy(*stream), d_in, d_in + num_items, typename std::iterator_traits<decltype(d_out)>::value_type{}), 1).wait();
```
Comment (zhimingwang36) [inline clang/test/dpct/query_api_mapping/CUB/cub_device.cu:79]: same as above.

File: clang/test/dpct/query_api_mapping/CUB/cub_device.cu [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+// CHECK_REDUCE-NEXT:   stream->fill(d_out, oneapi::dpl::reduce(oneapi::dpl::execution::device_policy(*stream), d_in, d_in + num_items, init_value, op), 1).wait();
+// CHECK_REDUCE-EMPTY:
+
+// RUN: dpct --cuda-include-path="%cuda-path/include" --stop-on-parse-err --query-api-mapping=cub::DeviceReduce::ReduceByKey | FileCheck %s -check-prefix=CHECK_REDUCE_BY_KEY
+// CHECK_REDUCE_BY_KEY: CUDA API:
+// CHECK_REDUCE_BY_KEY-NEXT:   cub::DeviceReduce::ReduceByKey(temp_storage/*void **/, temp_storage_bytes/*size_t &*/, d_keys_in/*KeysInputIteratorT*/, d_unique_out/*UniqueOutputIteratorT*/, d_values_in/*ValuesInputIteratorT*/, d_aggregates_out/*AggregatesOutputIteratorT*/, d_num_runs_out/*NumRunsOutputIteratorT*/, op/*ReductionOpT*/, num_items/*int*/, stream/*cudaStream_t*/);
+// CHECK_REDUCE_BY_KEY-NEXT: Is migrated to:
+// CHECK_REDUCE_BY_KEY-NEXT:   stream->fill(d_num_runs_out, std::distance(d_unique_out, oneapi::dpl::reduce_by_key(oneapi::dpl::execution::device_policy(*stream), d_keys_in, d_keys_in + num_items, d_values_in, d_unique_out, d_aggregates_out, std::equal_to<typename std::iterator_traits<decltype(d_keys_in)>::value_type>(), op).first), 1).wait();
```
Comment (zhimingwang36) [inline clang/test/dpct/query_api_mapping/CUB/cub_device.cu:68]: same as above.

File: clang/test/dpct/query_api_mapping/CUB/cub_device.cu (hunk 3 above)
Comment (yihwang-nv) [reply clang/test/dpct/query_api_mapping/CUB/cub_device.cu:35]: fixed

File: clang/test/dpct/query_api_mapping/CUB/cub_device.cu (hunk 4 above)
Comment (yihwang-nv) [reply clang/test/dpct/query_api_mapping/CUB/cub_device.cu:46]: fixed

File: clang/test/dpct/query_api_mapping/CUB/cub_device.cu (hunk 6 above)
Comment (yihwang-nv) [reply clang/test/dpct/query_api_mapping/CUB/cub_device.cu:68]: fixed

File: clang/test/dpct/query_api_mapping/CUB/cub_device.cu (hunk 5 above)
Comment (yihwang-nv) [reply clang/test/dpct/query_api_mapping/CUB/cub_device.cu:79]: fixed

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**